### PR TITLE
Refactor phpactor-update

### PR DESCRIPTION
**Pull request by zonuexe**
## Changes

* Rename `phpactor-update` to `phpactor-install-or-update`.
* Install Phpactor in `~/.emacs.d/phpactor/` instead of the directory where the package was installed (it is better to decouple those as package updates don't necessarily require an update of phpactor).

### CODE REVIEW

File: phpactor.el [hunk 1]
```diff
@@ -92,18 +116,28 @@
...
   "Install or update phpactor inside phpactor.el's folder."
   (interactive)
-  (let ((package-folder (phpactor--get-package-directory))
-        (composer-executable (car (composer--find-executable))))
-    (unless composer-executable (error "`composer' not found"))
-    (setq default-directory package-folder)
-    (call-process composer-executable nil (get-buffer-create phpactor--buffer-name) nil "install" "--no-dev")))
+  (let* ((default-directory (phpactor--get-package-directory))
+         (directory (or phpactor--base-directory
+                        phpactor--remote-composer-file-url-dir)))
```
Comment (kermorgant) [inline phpactor.el]: why (or in which case) would you look into `phpactor--remote-composer-file-url-dir` ?

File: phpactor.el [hunk 2]
```diff
@@ -92,18 +116,28 @@
           vendor-executable))
       (error "Phpactor not found.  Please run phpactor-update")))
 
-(defun phpactor-update ()
+;;;###autoload
+(defun phpactor-install-or-update ()
```
Comment (kermorgant) [inline phpactor.el]: :+1:

File: phpactor.el [hunk 3]
```diff
@@ -92,18 +116,28 @@
...
+  (let* ((default-directory (phpactor--get-package-directory))
+         (directory (or phpactor--base-directory
+                        phpactor--remote-composer-file-url-dir)))
+    (unless (file-directory-p phpactor-package-directory)
+      (make-directory phpactor-package-directory))
+    (cl-loop for file in '("composer.json" "composer.lock")
+             for code = (format "copy(%s, %s)"
+                                (php-runtime-quote-string (concat directory file))
+                                (php-runtime-quote-string (concat phpactor-package-directory file)))
+             do (message code)
```
Comment (kermorgant) [inline phpactor.el]: isn't that `message` some debug you forgot to remove ?

File: phpactor.el [hunk 4]
```diff
@@ -45,11 +45,28 @@
...
+;; Custom variables
+;;;###autoload
+(defgroup phpactor nil
+  "PHP refactoring and introspection"
+  :prefix "phpactor-"
+  :group 'tools
+  :group 'php)
+
+;;;###autoload
+(defcustom phpactor-package-directory
```
Comment (kermorgant) [inline phpactor.el]: hmm, I'm a bit confused by the `package` term. for me, package relates to phpactor.el, but it seems we're installing phpactor in this folder.

unless I'm wrong, maybe just `phpactor-directory` ?

File: phpactor.el [hunk 5]
```diff
@@ -76,12 +93,19 @@
 
 (defvar phpactor--buffer-name "*Phpactor*")
 
+;;; Constants
 (defconst phpactor-command-name "phpactor")
 (defconst phpactor--supported-rpc-version "1.0.0")
+(defconst phpactor--base-directory
```
Comment (kermorgant) [inline phpactor.el]: just to be sure I understand, this looks like the folder previously returned by `phpactor--get-package-directory` (ie something like ~/.emacs.d/elpa/phpactor-date). Am I correct ?

File: phpactor.el [hunk 6]
```diff
@@ -92,18 +116,28 @@
...
+                                (php-runtime-quote-string (concat phpactor-package-directory file)))
+             do (message code)
+             do (php-runtime-expr code))
+    (composer nil "install" "--no-dev")))
+(defalias 'phpactor-update #'phpactor-install-or-update)
 
 (defun phpactor--get-package-directory ()
   "Return the folder where phpactor.el is installed."
-  (file-name-directory(locate-library "phpactor.el")))
+  (or phpactor-package-directory
```
Comment (kermorgant) [inline phpactor.el]: following my previous comment about this function, maybe it is not needed ?

**Comment by kermorgant** [comment]
Hi @zonuexe 

I've just added a commit illustrating my view of the situation. It worked for me but maybe I miss some things you had in mind ?

**Comment by zonuexe** [comment]
@kermorgant I'm sorry, since this commit https://github.com/emacs-php/phpactor.el/pull/86/commits/2b2695ab14b666b85d9df5adc7608f158e479ead is not my intended code, I will delete it once in history.

**Comment by zonuexe** [comment]
@kermorgant 
I reorganized the commit.  Please point out again points that you do not know the intention of change.
Thank you.

**Comment by zonuexe** [comment]
@kermorgant Thank you for reviewing.

I merged this branch into develop.  I am not going to merge this commit for a while for a while, so please correct the incomprehensible places and naming incongruity.

**Comment by kermorgant** [comment]
@zonuexe I wonder, did you miss my pending comments about downloading composer.json etc. ?

**Comment by zonuexe** [comment]
@kermorgant Is it 2b2695ab14b666b85d9df5adc7608f158e479ead or anything else?

**Comment by kermorgant** [comment]
No, I meant those ones, about `phpactor--lisp-directory`

image

**Comment by zonuexe** [comment]
@kermorgant I see, the comment is in "pending" state, I can not read it.

**Comment by zonuexe** [comment]
## Why not composer install in the same directory as Lisp?

There are several reasons for this. In Emacs we can run Lisp files in several ways.
Lisp files are not only installed by package.el, they can be introduced by other mechanisms, or the Lisp code can be executed directly from the buffer.
That means there are several cases where directories can not be specified in either byte-compiled-dir or locate-directory. 
Or that may not have a directory at all in the first place. I would like to run it even under such circumstances.
In rare cases, it is possible that byte-compiled files are moved to other environments and the paths do not match.

Next, consider the case where it was installed by package.el or byte compiled normally.
In these cases we can assume that we can locate the appropriate `composer.json` and `composer.lock`.
However, since package.el contains the version number in the directory name, the user must set up Phpactor each time it is updated.
By install the files in a directory that is not dependent on the directory of the package, users can reduce the number of times to run `phpactor-install-or-update` frequently.

There are pros and cons in this implementation.  By having users `composer install` in the same location as the Lisp file every time you update, users always get the version of Phpactor recommended by `phpactor.el`.  In the current implementation, the user may continue to use the old Phpactor and cause trouble.

File: phpactor.el [hunk 7]
```diff
@@ -93,12 +93,19 @@
...
 (defvar phpactor--buffer-name "*Phpactor*")
 
+;;; Constants
 (defconst phpactor-command-name "phpactor")
 (defconst phpactor--supported-rpc-version "1.0.0")
+(defconst phpactor--base-directory
+  (eval-when-compile
+    (when (and (boundp 'byte-compile-current-file) byte-compile-current-file)
+      (file-name-directory byte-compile-current-file))))
+(defconst phpactor--remote-composer-file-url-dir
```
[the remark this answers is not in the document]
Comment (zonuexe) [reply phpactor.el]: @kermorgant 
Yes, we can not get byte-compiled-dir in the following cases:

- Lisp code was directly evaluated on the buffer.
- Not byte-compiled Lisp file was loaded

We do not see these symptoms when installing the Lisp package from `package.el`, but in that case occasionally.

**Comment by kermorgant** [comment]
@zonuexe Thanks for taking the time to explain your reasoning. 

I'm fine with the last part, about reducing `the number of times to run `phpactor-install-or-update` frequently`. The risk about using an old version of Phpactor is something we should be able to cope with elegantly with Phpactor's RPC versioning.

Then about the first part... First, I should stress once again I'm quite inexperienced with emacs so my view on this is necesserily narrowed. But from a practical perspective, I still can't see why a user would install a package differently than with package.el or cloning the git repo directly.

This is not a big deal per se, I'm fine with this but as I was rewiewing this, I tried to get into the details ;-)
And I guess I insisted a bit on this because I might have an opposite view on such things : I'd personnally be more of a dictator and enforce a standard way of installing this package and by that avoid some corner cases. But again, it's not a problem, it's just to share a bit how I view things.

**Comment by zonuexe** [comment]
@kermorgant 
> And I guess I insisted a bit on this because I might have an opposite view on such things : I'd personnally be more of a dictator and enforce a standard way of installing this package and by that avoid some corner cases. But again, it's not a problem, it's just to share a bit how I view things.

That idea makes sense.  Still do not force all users to install from the package.  For example, this implementation could be replaced by a message saying “Please set `phpactor-executable` or `phpactor-lisp-directory`.”

The balance between "saving user's labor" and "keeping the code simple" is a difficult decision.  In this implementation, I emphasized the user's convenience. When I need to maintain this code several times in the future, I agree to remove it.

**Pull request merged**